### ISSUE TITLE
Implementing Equals()/GetHashCode() causes YesSql to fail

### ISSUE THREAD

**Issue by Jabbersii** [sebastienros/yessql#167]
Implementing Equals() and GetHashCode() on an entity class causes Save to fail.

With a model like this:
```csharp
class Person
{
    public int Id { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }

    public override bool Equals(object obj)
    {
        var other = obj as Person;
        return other != null &&
            this.FirstName == other.FirstName &&
            this.LastName == other.LastName;
    }

    // Any implementation of GetHashCode - using this for example
    public override in GetHashCode() => HashCode.Combine(FirstName, LastName);
}
```

A test like this will fail:
``` csharp
IStore store = InitStore(); // Create a store with any configuration
using (var session = store.CreateSession())
{
    var person = new Person() { FirstName = "Joe", LastName = "Bloggs" };
    session.Save(person);
}

using (var session = store.CreateStore())
{
    var dbPerson = session.Query().For<Person>().FirstOrDefault();
    var dbPerson.FirstName = "Bob";
    session.Save(dbPerson); // Fails on save
}
```

With the code above, `Session.Save(object)` throws the error: `System.ArgumentException : An item with the same key has already been added. Key: 1`

If `Id` is included in the `Equals()`/`GetHashCode()` methods, the test fails on the *first* save with the exception: `System.InvalidOperationException : The object to save was not found in identity map.`

I do have a change which can fix this, but I wasn't sure about raising a pull request.

**Comment by sebastienros** [comment]
Probably because you have two items with the same hashcode (i.e. same Firstname/LAstname), however I would expect it to handle conflicts. 

The issue with the Id is the opposite one, it varies the hashcode before and after you call Save, so it won't find it in the expected dictionary bucket.

**Comment by AkosLukacs** [comment]
Basically any implementation of `GetHashCode` makes YesSql unusable.
Any idea of a solution / workaround?

**Comment by dlidstrom** [comment]
@AkosLukacs use a wrapper class instead, i.e. keep the db entities free of any logic, including equality comparison.

**Comment by AkosLukacs** [comment]
> @AkosLukacs use a wrapper class instead, i.e. keep the db entities free of any logic, including equality comparison.

@dlidstrom Yes, did that. Btw the type was a record type from F#, the compiler "implements" GetHashCode.

**Comment by dlidstrom** [comment]
Good catch! That means F# records should be annotated with [ReferenceEquality](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-referenceequalityattribute.html) attribute in order to be used with YesSql.
